packages/timeline-state-resolver/src/service/commandExecutor.ts

```diff
@@ -18,11 +22,44 @@ export class CommandExecutor<DeviceState, Command extends CommandWithContext> {
 		commands.sort((a, b) => (b.preliminary ?? 0) - (a.preliminary ?? 0))
 		const totalTime = commands[0].preliminary ?? 0
 
-		if (this.mode === 'salvo') {
-			return this._executeCommandsSalvo(totalTime, commands, measurement)
-		} else {
-			return this._executeCommandsSequential(totalTime, commands, measurement)
+		const salvoCommands: Command[] = []
+		const sequentialCommands: Command[] = []
+
+		for (const command of commands) {
+			const mode = command.executeMode || this.defaultMode
+			if (mode === ExecuteMode.SEQUENTIAL) {
+				sequentialCommands.push(command)
+			} else {
+				salvoCommands.push(command)
+			}
 		}
+
+		// The execution logic is as follows:
+		// * When the commands sent into executeCommands() is called a BATCH.
+		//
+		// * Salvos should wait for Salvos from previous BATCH to have finished before executing
+		// * Salvos within the same BATCH are executed in parallel.
+		//
+		// * Sequentials should wait for all Salvos and Sequentials from previous BATCH to have finished before executing
+		// * Sequentials should be executed after all Salvos from the same BATCH have been executed
+		// * Sequentials within the same BATCH are executed in order
```

Comment:
I am somewhat worried about this algorithm mixing commands from different batches. After all, Salvos from Batch N+1 have been generated assuming that Salvos _and_ Sequentials from Batch N have already been sent. I feel like each batch should be done, before we start sending commands for the next batch? They are effectively applying state, and thus each batch depends on the previous one to have been applied?

Reply:
Perhaps a solution to this is to write some really good documentation about the intention of the type of commands?

For my use case, I would need it to work this way, because:

* Batch 1:
  * Sequntial: Quantel stop command (delayed, so will wait and not resolve until after an `await sleep(x)` )
* Batch 2:
  * Salvo: cancel any waiting Quantel operations
  * Sequntial: Quantel play command

So in this case, the cancel-salvo _must not_ wait for the previous batch to finish.

---

The only other possible solution to this would be to handle specific situations like this internally in the device - but that goes agains the idea of the rewrite of the TSR devices where we try to unify the integration..